planner/core/rule_join_reorder.go

```diff
@@ -34,7 +34,7 @@ import (
 // For example: "InnerJoin(InnerJoin(a, b), LeftJoin(c, d))"
 // results in a join group {a, b, c, d}.
 func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction,
-	otherConds []expression.Expression, joinTypes []JoinType, hintInfo []*tableHintInfo, hasOuterJoin bool) {
+	otherConds []expression.Expression, joinTypes []*abundantJoinType, hintInfo []*tableHintInfo, hasOuterJoin bool) {
```

Comment:
Is it possible to add another param for `outerBindCondition` instead of using  so-called `abundantJoinType`, which is not easy to follow.

Reply:
the outer condition should be exactly mapped to a corresponding join type. it will be like following if did
joinType    :  [inner][inner][outer]
bindCond  : [nil][nil][some cond]

Reply:
Yep, I mean the outerBindCondition is also an array or list.

Reply:
@fixdb what about just renaming the `abundantJoinType` to something like `joinTypeWithExtMsg`?

Reply:
Your call, as long as it is not confusing.